File: Python3/249. Group Shifted Strings.py
class Solution:
    def groupStrings(self, strings: List[str]) -> List[List[str]]:
        patterns = collections.defaultdict(list)
        
        # each char minus a
        a = ord('a')
        
        for string in strings:
            
            # we'll use this diff to compute the base pattern of this kind of string
            diff = ord(string[0]) - a
            
            temp = ''
            #print(diff)
            for char in string:
                temp += chr((ord(char)-diff)%26)
                
            patterns[temp].append(string)
            
        return list(patterns.values())

File: Java/249. Group Shifted Strings.java
class Solution {
    public List<List<String>> groupStrings(String[] strings) {
        Map<String, List<String>> patterns = new HashMap<>();
        
        for(String s: strings){
            String key = getHash(s);
            if(patterns.get(key) == null){
                patterns.put(key, new ArrayList<>());
            }
            patterns.get(key).add(s);
        }
        
        // transfer the values to the result list
        List<List<String>> res = new ArrayList<>();
        
        for(List<String> s: patterns.values()){
            res.add(s);
        }
        
        return res;
    }
    
    private String getHash(String s){
        char[] chars = s.toCharArray();
        
        // shifts that needed to transform the first char into 'a'
        int diff = chars[0];
        
        for(int i=0; i<chars.length; i++){
            chars[i] = shifter(chars[i], diff);
        }
        
        String key = String.valueOf(chars);
        return key;
    }
    
    private char shifter(char c, int diff){
        return (char) ((c-diff+26)%26 + 'a');
    }
}